### PR TITLE
Refactors MutableSpan -> zipkin2.Span into its own type

### PR DESCRIPTION
**Pull request by adriancole**
This furthers decoupling by isolating most zipkin2 conversion code into
`MutableSpanConverter`. Rather than relying on package private stuff,
this adds `MutableSpan.forEachTag` to let you copy data without
allocating an iterator or closing over the target object (as is the case
with `Map.forEach`.

### CODE REVIEW

File: brave/src/main/java/brave/internal/recorder/MutableSpan.java [hunk 1]
```diff
@@ -31,11 +38,27 @@
    * To reduce the amount of allocation, collocate annotations with tags in a pair-indexed list.
    * This will be (startTimestamp, value) for annotations and (key, value) for tags.
    */
-  final List<Object> pairs = new ArrayList<>(6); // assume 3 tags and no annotations
+  final ArrayList<Object> pairs;
```
Comment (wu-sheng) [inline brave/src/main/java/brave/internal/recorder/MutableSpan.java]: `ArrayList` will cost when the length extend happens. Linkedlist is always faster.

File: brave/src/main/java/brave/internal/recorder/MutableSpan.java [hunk 2]
```diff
@@ -117,6 +165,27 @@ public void tag(String key, String value) {
...
+      Object first = pairs.get(i);
+      if (first instanceof Long) continue; // hit an annotation
+      tagConsumer.accept(target, first.toString(), pairs.get(i + 1).toString());
+    }
+  }
+
+  public <T> void forEachAnnotation(AnnotationConsumer<T> annotationConsumer, T target) {
+    for (int i = 0, length = pairs.size(); i < length; i += 2) {
+      Object first = pairs.get(i);
+      if (first instanceof String) continue; // hit a tag
```
Comment (wu-sheng) [inline brave/src/main/java/brave/internal/recorder/MutableSpan.java]: `instanceof` has some performance concern when you are using jre6(jre7 maybe)

File: brave/src/main/java/brave/internal/recorder/MutableSpan.java (hunk 1 above)
Comment (adriancole) [reply brave/src/main/java/brave/internal/recorder/MutableSpan.java]: https://github.com/google/error-prone/blob/0cc518144272ea6150c85e2cfb937c82a8d85223/docs/bugpattern/JdkObsolete.md#linkedlist so we will have to make an exception if benchmarks show you are correct. this is cool though.. internal detail will check it!

File: brave/src/main/java/brave/internal/recorder/MutableSpan.java (hunk 2 above)
Comment (adriancole) [reply brave/src/main/java/brave/internal/recorder/MutableSpan.java]: heh well it seems the reverse is true starting at java 8 (instanceof vs getClass ==)
https://stackoverflow.com/questions/103564/the-performance-impact-of-using-instanceof-in-java

Comment (wu-sheng) [reply brave/src/main/java/brave/internal/recorder/MutableSpan.java]: Ha, look like something is still right in the memory.

File: brave/src/main/java/brave/internal/recorder/MutableSpan.java (hunk 1 above)
Comment (wu-sheng) [reply brave/src/main/java/brave/internal/recorder/MutableSpan.java]: I am always wondering, why linkedlist should not be used.

**Comment by adriancole** [comment]
Thanks for the comments. For now, sticking with ArrayList. If in the future we have a problem, we can always do something like @felixbarny with recycling the mutable spans (which would re-use the arraylists).

Here's why we are not going to change from LinkedList to ArrayList (besides errorprone saying don't). Later, we can revisit, but anyway it is quite quick right now and not better off even when the arraylist grows (client span grows the array) vs linkedlist.

```
ArrayList
===========================================================================
Benchmark                                 Mode  Cnt   Score   Error   Units
MutableSpanBenchmarks.makeBigClientSpan  thrpt   15   3.749 ± 0.034  ops/us
MutableSpanBenchmarks.makeServerSpan     thrpt   15  22.828 ± 0.447  ops/us

Benchmark                                             Mode  Cnt  Score   Error   Units
MutableSpanConverterBenchmarks.convertBigClientSpan  thrpt   15  2.494 ± 0.099  ops/us
MutableSpanConverterBenchmarks.convertServerSpan     thrpt   15  3.617 ± 0.098  ops/us


LinkedList
===========================================================================
Benchmark                                 Mode  Cnt   Score   Error   Units
MutableSpanBenchmarks.makeBigClientSpan  thrpt   15   2.830 ± 0.027  ops/us
MutableSpanBenchmarks.makeServerSpan     thrpt   15  18.322 ± 0.181  ops/us

Benchmark                                             Mode  Cnt  Score   Error   Units
MutableSpanConverterBenchmarks.convertBigClientSpan  thrpt   15  2.216 ± 0.051  ops/us
MutableSpanConverterBenchmarks.convertServerSpan     thrpt   15  3.510 ± 0.091  ops/us
```

**Comment by adriancole** [comment]
PS definitely instanceof check when removed allowed perf much better, so thanks about that @wu-sheng!

**Comment by felixbarny** [comment]
To get a broader picture of the performance, I suggest using `@BenchmarkMode(Mode.SampleTime)` which gives you percentiles for the response time. Also looking at the allocation rate with `-profiler gc` can yield interesting insights.

File: brave/src/main/java/brave/internal/recorder/MutableSpan.java (hunk 1 above)
Comment (felixbarny) [reply brave/src/main/java/brave/internal/recorder/MutableSpan.java]: The main criticism of LinkedLists is pointer chasing. When using an ArrayList, the memory layout is much denser and the memory access pattern is easier to predict by CPUs. With linked lists, the next elements are less likely in the same cache line which leads to expensive cache misses and main memory access.

**Comment by adriancole** [comment]
OK thanks for the tips. We will do a different PR to make the objects recyclable, but for now the tests still show favor to arraylist. Ideally arraylist becomes even better once we recycle. Also, there's a lot of coding left to do, so don't want to over optimize when we know we will recycle later (which will invalidate a lot of our assumptions!)


ArrayList
```
Benchmark                                                                   Mode     Cnt     Score     Error   Units
MutableSpanBenchmarks.makeBigClientSpan                                   sample  416604     0.366 ±   0.020   us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.00           sample             0.276             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.50           sample             0.302             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.90           sample             0.323             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.95           sample             0.336             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.99           sample             0.633             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.999          sample            10.864             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.9999         sample            31.725             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p1.00           sample           737.280             us/op
MutableSpanBenchmarks.makeBigClientSpan:·gc.alloc.rate                    sample      15   732.102 ±  11.394  MB/sec
MutableSpanBenchmarks.makeBigClientSpan:·gc.alloc.rate.norm               sample      15   328.068 ±   0.006    B/op
MutableSpanBenchmarks.makeBigClientSpan:·gc.churn.PS_Eden_Space           sample      15   737.155 ±  28.188  MB/sec
MutableSpanBenchmarks.makeBigClientSpan:·gc.churn.PS_Eden_Space.norm      sample      15   330.325 ±  11.415    B/op
MutableSpanBenchmarks.makeBigClientSpan:·gc.churn.PS_Survivor_Space       sample      15     0.234 ±   0.064  MB/sec
MutableSpanBenchmarks.makeBigClientSpan:·gc.churn.PS_Survivor_Space.norm  sample      15     0.105 ±   0.029    B/op
MutableSpanBenchmarks.makeBigClientSpan:·gc.count                         sample      15   266.000            counts
MutableSpanBenchmarks.makeBigClientSpan:·gc.time                          sample      15   131.000                ms
MutableSpanBenchmarks.makeServerSpan                                      sample  413185     0.104 ±   0.011   us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.00                 sample             0.026             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.50                 sample             0.082             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.90                 sample             0.092             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.95                 sample             0.100             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.99                 sample             0.177             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.999                sample             8.538             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.9999               sample            13.949             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p1.00                 sample           707.584             us/op
MutableSpanBenchmarks.makeServerSpan:·gc.alloc.rate                       sample      15  1850.922 ±  91.591  MB/sec
MutableSpanBenchmarks.makeServerSpan:·gc.alloc.rate.norm                  sample      15   144.009 ±   0.001    B/op
MutableSpanBenchmarks.makeServerSpan:·gc.churn.PS_Eden_Space              sample      15  1848.837 ± 109.138  MB/sec
MutableSpanBenchmarks.makeServerSpan:·gc.churn.PS_Eden_Space.norm         sample      15   143.851 ±   4.964    B/op
MutableSpanBenchmarks.makeServerSpan:·gc.churn.PS_Survivor_Space          sample      15     0.229 ±   0.036  MB/sec
MutableSpanBenchmarks.makeServerSpan:·gc.churn.PS_Survivor_Space.norm     sample      15     0.018 ±   0.003    B/op
MutableSpanBenchmarks.makeServerSpan:·gc.count                            sample      15   294.000            counts
MutableSpanBenchmarks.makeServerSpan:·gc.time                             sample      15   145.000                ms
```

LinkedList
```
Benchmark                                                                   Mode     Cnt     Score     Error   Units
MutableSpanBenchmarks.makeBigClientSpan                                   sample  342856     0.555 ±   0.032   us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.00           sample             0.341             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.50           sample             0.399             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.90           sample             0.481             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.95           sample             0.764             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.99           sample             1.230             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.999          sample            15.568             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p0.9999         sample            84.224             us/op
MutableSpanBenchmarks.makeBigClientSpan:makeBigClientSpan·p1.00           sample          1507.328             us/op
MutableSpanBenchmarks.makeBigClientSpan:·gc.alloc.rate                    sample      15   907.022 ± 113.815  MB/sec
MutableSpanBenchmarks.makeBigClientSpan:·gc.alloc.rate.norm               sample      15   576.118 ±   0.039    B/op
MutableSpanBenchmarks.makeBigClientSpan:·gc.churn.PS_Eden_Space           sample      15   906.373 ± 133.563  MB/sec
MutableSpanBenchmarks.makeBigClientSpan:·gc.churn.PS_Eden_Space.norm      sample      15   574.580 ±  32.393    B/op
MutableSpanBenchmarks.makeBigClientSpan:·gc.churn.PS_Survivor_Space       sample      15     0.238 ±   0.059  MB/sec
MutableSpanBenchmarks.makeBigClientSpan:·gc.churn.PS_Survivor_Space.norm  sample      15     0.154 ±   0.046    B/op
MutableSpanBenchmarks.makeBigClientSpan:·gc.count                         sample      15   230.000            counts
MutableSpanBenchmarks.makeBigClientSpan:·gc.time                          sample      15   131.000                ms
MutableSpanBenchmarks.makeServerSpan                                      sample  477195     0.119 ±   0.016   us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.00                 sample             0.065             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.50                 sample             0.081             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.90                 sample             0.121             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.95                 sample             0.166             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.99                 sample             0.268             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.999                sample             8.736             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p0.9999               sample            25.287             us/op
MutableSpanBenchmarks.makeServerSpan:makeServerSpan·p1.00                 sample          2125.824             us/op
MutableSpanBenchmarks.makeServerSpan:·gc.alloc.rate                       sample      15  2861.716 ± 483.178  MB/sec
MutableSpanBenchmarks.makeServerSpan:·gc.alloc.rate.norm                  sample      15   288.014 ±   0.008    B/op
MutableSpanBenchmarks.makeServerSpan:·gc.churn.PS_Eden_Space              sample      15  2877.346 ± 486.107  MB/sec
MutableSpanBenchmarks.makeServerSpan:·gc.churn.PS_Eden_Space.norm         sample      15   289.829 ±   9.566    B/op
MutableSpanBenchmarks.makeServerSpan:·gc.churn.PS_Survivor_Space          sample      15     0.249 ±   0.083  MB/sec
MutableSpanBenchmarks.makeServerSpan:·gc.churn.PS_Survivor_Space.norm     sample      15     0.025 ±   0.007    B/op
MutableSpanBenchmarks.makeServerSpan:·gc.count                            sample      15   260.000            counts
MutableSpanBenchmarks.makeServerSpan:·gc.time                             sample      15   137.000                ms
```

**Pull request merged**